src/gtfs/repository/CIFRepository.ts

```diff
@@ -79,51 +85,52 @@ export class CIFRepository {
-SUBSTRING(s.valid_days, 2, 1 ) as tuesday,
-SUBSTRING(s.valid_days, 3, 1 ) as wednesday,
-SUBSTRING(s.valid_days, 4, 1 ) as thursday,
-SUBSTRING(s.valid_days, 5, 1 ) as friday,
-SUBSTRING(s.valid_days, 6, 1 ) as saturday,
-SUBSTRING(s.valid_days, 7, 1 ) as sunday,
-
-loc.crs_code as crs_code, s.stp_indicator as stp_indicator,
-sloc.public_arrival_time, sloc.public_departure_time,
-IF(s.train_status="S", "SS", s.train_category) AS train_category, 
-IFNULL(sloc.scheduled_arrival_time, sloc.scheduled_pass_time) AS scheduled_arrival_time, 
-IFNULL(sloc.scheduled_departure_time, sloc.scheduled_pass_time) AS scheduled_departure_time,
-sloc.platform, e.atoc_code, sloc.schedule_location_id AS stop_id, 
-COALESCE(sloc.activity, "") as activity, s.reservations, s.train_class
+    const sqlQuery = `
+    SELECT 
+    s.schedule_id as id,
+    s.train_uid, 
+    e.rsid as retail_train_id, 
+    greatest(s.wef_date, COALESCE(s.import_wef_date, s.wef_date)) as runs_from, 
+    least(s.weu_date, COALESCE(s.import_weu_date, s.weu_date)) as runs_to,
 
-FROM cif_schedule as s 
+    SUBSTRING(s.valid_days, 1, 1 ) as monday,
+    SUBSTRING(s.valid_days, 2, 1 ) as tuesday,
+    SUBSTRING(s.valid_days, 3, 1 ) as wednesday,
+    SUBSTRING(s.valid_days, 4, 1 ) as thursday,
+    SUBSTRING(s.valid_days, 5, 1 ) as friday,
+    SUBSTRING(s.valid_days, 6, 1 ) as saturday,
+    SUBSTRING(s.valid_days, 7, 1 ) as sunday,
 
-LEFT JOIN cif_schedule_extra as e
-	ON e.schedule_id = s.schedule_id
-LEFT JOIN cif_schedule_location as sloc
-	ON sloc.schedule_id = s.schedule_id
-LEFT JOIN master_location as loc
-	ON sloc.tiploc = loc.tiploc
-	
-WHERE 
-	(sloc.schedule_location_id IS NULL OR (loc.crs_code IS NOT NULL AND loc.crs_code != "") )
-	AND s.wef_date < ?
-  AND s.weu_date >= ?
-  AND (s.import_weu_date IS NULL OR (s.import_weu_date > ?) )
-  
-  HAVING runs_to >= runs_from
-ORDER BY stp_indicator DESC, s.schedule_id, sloc.location_order
+    loc.crs_code as crs_code, s.stp_indicator as stp_indicator,
+    sloc.public_arrival_time, sloc.public_departure_time,
+    IF(s.train_status="S", "SS", s.train_category) AS train_category, 
+    IFNULL(sloc.scheduled_arrival_time, sloc.scheduled_pass_time) AS scheduled_arrival_time, 
+    IFNULL(sloc.scheduled_departure_time, sloc.scheduled_pass_time) AS scheduled_departure_time,
+    sloc.platform, e.atoc_code, sloc.schedule_location_id AS stop_id, 
+    COALESCE(sloc.activity, "") as activity, s.reservations, s.train_class
 
+    FROM cif_schedule as s 
 
-      `, [this.endRange, this.startRange, this.startRange]);
+    LEFT JOIN cif_schedule_extra as e
+      ON e.schedule_id = s.schedule_id
+    LEFT JOIN cif_schedule_location as sloc
+      ON sloc.schedule_id = s.schedule_id
+    LEFT JOIN master_location as loc
+      ON sloc.tiploc = loc.tiploc
+      
+    WHERE 
+      (sloc.schedule_location_id IS NULL OR (loc.crs_code IS NOT NULL AND loc.crs_code != "") )
+      AND s.wef_date < '${this.endRange}'
```

Comment:
This query shouldn't be a template string, as this will expose us to SQL injection. If I passed through a startRange query of `; DROP TABLE "some important table`, this will currently execute it happily.

We should still use a prepared statement, then populate the the variables. Perhaps have this statement `
AND (s.schedule_type != 'VSTP' OR ?)` where that `?` would be the allow VSTP flag.

Reply:
Totally ignored the security side..  
pre-define the condition and use a boolean flag to control the condition is a nice way, I have adapt this method in the new code in: 2c73484

Reply:
Looks good!